src/lib/projectConfigValidation.ts

```diff
@@ -0,0 +1,439 @@
+  workspace.onDidChangeTextDocument(
+    debounce(
+      (event) => validateDocument(event.document),
+      VALIDATION_DEBOUNCE_TIME
+    )
+  );
+  workspace.onDidSaveTextDocument(validateDocument);
+  workspace.onDidCloseTextDocument((doc) =>
+    diagnosticsCollection.delete(doc.uri)
+  );
+  workspace.textDocuments.forEach(validateDocument);
+}
+
+async function validateDocument(document: TextDocument): Promise<void> {
+  try {
+    const diagnostics = await validateDocumentDiagnostics(document);
+    diagnosticsCollection.set(document.uri, diagnostics);
+  } catch (error) {
+    console.error('Error validating document:', error);
+  }
+}
+
+async function validateDocumentDiagnostics(
+  document: TextDocument
+): Promise<Diagnostic[]> {
+  if (!document.fileName.endsWith('-hsmeta.json')) return [];
+  const projectRoot = await findProjectRoot(document.uri);
+  if (!projectRoot) return [];
+  const projectVersion = await getProjectVersion(projectRoot);
+  if (!projectVersion) return [];
+  // Any file that fails the above checks is one that we don't handle
+  let metaFile: any; // The type is whatever JSON the user provided us with
+  try {
+    metaFile = JSON.parse(document.getText());
+  } catch (error) {
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        getInvalidJsonError(),
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  const root = jsonc.parseTree(document.getText());
+  if (!root) {
+    // Should not happen, we've already parsed it above without jsonc
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        getInvalidJsonError(),
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  if (!('type' in metaFile)) {
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        getMissingTypeError(),
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  const uidError = validateUid(metaFile.uid);
+  if (uidError) {
+    const uidNode = jsonc.findNodeAtLocation(root, ['uid']);
+    return [
+      new Diagnostic(
+        uidNode
+          ? new Range(
+              document.positionAt(uidNode.offset),
+              document.positionAt(uidNode.offset + uidNode.length)
+            )
+          : new Range(0, 0, 0, 0),
+        uidError,
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  if (!('config' in metaFile)) {
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        getMissingConfigError(),
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  const accountId = getAccountId();
+  if (!accountId) {
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        `${getMissingAccountIdError()}. Authenticate account with the HubSpot CLI`,
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  const validator = await getCachedValidatorForExternalType(
+    projectRoot,
+    projectVersion,
+    accountId,
+    metaFile.type
+  );
+  if ('error' in validator) {
```

Comment:
I don't think we should use the in operator here since it doesn't handle nullable cases well